### ISSUE TITLE
Remove no-default-alpn

### ISSUE THREAD

**Issue by martinthomson** [MikeBishop/dns-alt-svc#193]
This is of no lasting value and it only represents a marginal gain in space for a non-trivial gain in complexity.  Having an ALPN field in every HTTPS record just isn't that many bytes (2+2+3=7 for just "h2", 2+2+3+9=16 for "h2,http/1.1").

I've raised this a couple of times in the past, but I never received a satisfactory response.

**Comment by bemasc** [comment]
The goal here was not efficiency of encoding, it was to avoid a certain category of administrative mistakes.  Specifically, without this in place, it would be easy for a zone owner to add a single record with "alpn=h3 echconfig=...", test locally, see everything is working, and not realize that they just disabled their website for ~10% of users (who don't have UDP routability).

This represents something of a compromise between participants who wanted to be able to run QUIC and TCP on separate endpoints (or even QUIC-only services in unusual situations), and participants who wanted HTTPS-over-TCP to be required for every endpoint.

**Comment by martinthomson** [comment]
I don't find that sort of defensive design useful relative to the complexity it adds.  And it makes the configuration less transparent.

**Comment by bemasc** [comment]
The `no-default-alpn` business is the result of a Very Long Debate that I would prefer not to reopen, but it's not something anyone is particularly happy with.  Currently, getting rid of it would result in records being unusable in the default state.  To fix this, we would have to give `alpn` a default value of `http/1.1` (for the HTTPS RR) if the key is absent.

In the design you're proposing, `alpn=h2` means that you dropped support for HTTP/1.1-only clients.  (I believe even Chrome is sometimes HTTP/1.1-only, e.g. for WebSockets).  That strikes me as more likely to result in swearing sysadmins than the current design, and same for the `alpn=h3` case.

To put it another way, I think there are likely to be many orders of magnitude more zone owners configuring these RRs than software developers implementing them, so we should prioritize the ease of (working) configuration over the ease of implementation.

**Comment by tfpauly** [comment]
Would it be helpful to adjust the name of this parameter to make the effect more clear? For example, rather than `no-default-alpn`, you could call it `strict-alpn`, `restricted-alpn`, or something similar.

Making claims about what the default ALPN is might be causing some of the annoyance here. Instead, if we think of the list of ALPN values in the record as purely additive from the client’s perspective (always just a hint to add capability and never restrict), this key becomes a way to disable ‘additive’ mode and tell clients that the list is actually a hint of the only ALPNs that are supported.

**Comment by enygren** [comment]
Calling it strict-alpn or restricted-alpn would not be compatible with the changes in #202.  Are there other names that might be less confusing?

**Comment by tfpauly** [comment]
In what way would those proposals not be compatible?

**Comment by bemasc** [comment]
I think that should be #201.

"Strict" or "restricted" would seem to indicate that the client must not send anything other than listed ALPNs, or that the server won't accept anything else.  Neither of those readings match the intent of the ALPN selection process, in the current draft or in #201.

**Comment by tfpauly** [comment]
#201 makes more sense =)

Yes, I agree that it's not "strict". It's mainly a hint that the old ALPNs won't work.

What is the client behavior intention here? As a client, I'm honestly just ignoring it right now. I guess it's a way to say, if you think you can get rid of h1/h2, then delay trying those?

**Comment by enygren** [comment]
Example use-case would be having an RR set with two HTTPS records: one supports h1 and h2 and h3 and the other supports only h3.  The one with only h3 would have higher priority.  This would be to allow different service addresses (or different server pools) that are h3-only.  That may come in-hands as h3 opens up some interesting deployment possibilities.

**Comment by enygren** [comment]
Based on discussions, are there concerns about closing out this issue?
(While not perfect, this still seems like a reasonable compromise.)

**Comment by martinthomson** [comment]
"No" is hardly a compromise, but I can live with the outcome.
